pythonforandroid/graph.py

```diff
@@ -7,49 +7,79 @@
 from pythonforandroid.util import BuildInterruptingException
 
 
-class RecipeOrder(dict):
+def fix_deplist(deps):
+    """ Turn a dependency list into lowercase, and make sure all entries
+        that are just a string become a tuple of strings
+    """
+    deps = [
+        ((dep.lower(),) if not isinstance(dep, (list, tuple))
+         else tuple([dep_entry.lower() for dep_entry in dep]))
+        for dep in deps
+    ]
+    return deps
+
 
+class RecipeOrder(dict):
     def __init__(self, ctx):
         self.ctx = ctx
 
-    def conflicts(self, name):
+    def conflicts(self):
         for name in self.keys():
             try:
                 recipe = Recipe.get_recipe(name, self.ctx)
-                conflicts = recipe.conflicts
-            except IOError:
+                conflicts = [dep.lower() for dep in recipe.conflicts]
+            except ValueError:
                 conflicts = []
 
             if any([c in self for c in conflicts]):
                 return True
         return False
 
 
-def recursively_collect_orders(name, ctx, all_inputs, orders=[]):
+def get_dependency_tuple_list_for_recipe(recipe, blacklist=[]):
```

Comment:
Using mutable default parameters is discouraged in Python https://docs.python-guide.org/writing/gotchas/
Could you double check we're safe here and other places? As a rule of thumb I would just recommend to eventually use e.g. `blacklist=None`

Reply:
I'd say it's more than discouraged. Please don't ever use `[]` as a default argument.

See, e.g. : https://docs.python-guide.org/writing/gotchas/

Reply:
@AndreMiras @hackalog the parameter is never changed, it is a read-only parameter everywhere except for the initial `get_recipe_order_and_bootstrap` where I lowercase it at the top. `blacklist=None` would need me to add conditionals to all the subfunctions. I don't mind if that is really what you want, but I don't think that is actually a better solution and would look rather ugly

Reply:
(also compare `orders=[]` which was already there, where the same applies)

Reply:
let me know what your final decision is. should I still change it?

Reply:
If you definitely think your option is better and will not generate headache later, then let's leave it. I'll just pray for nobody to do something silly in here :pray: